operators/pkg/controller/common/certificates/ca_reconcile.go

```diff
@@ -0,0 +1,213 @@
+		return renewCA(cl, namer, owner, caCertValidity, scheme, caType)
+	}
+
+	// build CA
+	ca := buildCAFromSecret(caInternalSecret)
+	if ca == nil {
+		log.Info("Cannot build CA from secret, creating a new one", "owner", ownerNsn, "ca_type", caType)
+		return renewCA(cl, namer, owner, caCertValidity, scheme, caType)
+	}
+
+	// renew if cannot reuse
+	if !canReuseCA(ca, expirationSafetyMargin) {
+		log.Info("Cannot reuse existing CA, creating a new one", "owner", ownerNsn, "ca_type", caType)
+		return renewCA(cl, namer, owner, caCertValidity, scheme, caType)
+	}
+
+	// reuse existing CA
+	log.V(1).Info("Reusing existing CA", "owner", ownerNsn, "ca_type", caType)
+	return ca, nil
+}
+
+// renewCA creates and stores a new CA to replace one that might exist
+func renewCA(
+	client k8s.Client,
+	namer name.Namer,
+	owner v1.Object,
+	expireIn time.Duration,
+	scheme *runtime.Scheme,
+	caType CAType,
+) (*CA, error) {
+	ca, err := NewSelfSignedCA(CABuilderOptions{
+		Subject: pkix.Name{
+			CommonName:         string(caType) + "-" + rand.String(16),
+			OrganizationalUnit: []string{owner.GetName()},
+		},
+		ExpireIn: &expireIn,
+	})
+	if err != nil {
+		return nil, err
+	}
+	caInternalSecret := internalSecretForCA(ca, namer, owner, caType)
+
+	// create or update internal secret
+	reconciledCAInternalSecret := corev1.Secret{}
+	if err := reconciler.ReconcileResource(reconciler.Params{
+		Client:           client,
+		Expected:         &caInternalSecret,
+		NeedsUpdate:      func() bool { return true },
+		Owner:            owner,
+		Reconciled:       &reconciledCAInternalSecret,
+		Scheme:           scheme,
+		UpdateReconciled: func() { reconciledCAInternalSecret.Data = caInternalSecret.Data },
+	}); err != nil {
+		return nil, err
+	}
+
+	return ca, nil
+}
+
+// canReuseCA returns true if the given CA is valid for reuse
+func canReuseCA(ca *CA, expirationSafetyMargin time.Duration) bool {
+	return PrivateMatchesPublicKey(ca.Cert.PublicKey, *ca.PrivateKey) && certIsValid(*ca.Cert, expirationSafetyMargin)
+}
+
+// certIsValid returns true if the given cert is valid,
+// according to a safety time margin.
+func certIsValid(cert x509.Certificate, expirationSafetyMargin time.Duration) bool {
+	now := time.Now()
+	if now.Before(cert.NotBefore) {
+		log.Info("CA cert is not valid yet, will create a new one")
+		return false
+	}
+	if now.After(cert.NotAfter.Add(-expirationSafetyMargin)) {
+		log.Info("CA cert expired or soon to expire, will create a new one", "expiration", cert.NotAfter)
+		return false
+	}
+	return true
+}
+
+// internalSecretForCA returns a new internal Secret for the given CA.
+func internalSecretForCA(
+	ca *CA,
+	namer name.Namer,
+	owner v1.Object,
+	caType CAType,
+) corev1.Secret {
+	return corev1.Secret{
+		ObjectMeta: v1.ObjectMeta{
+			Namespace: owner.GetNamespace(),
+			Name:      CAInternalSecretName(namer, owner.GetName(), caType),
```

Comment:
Should we set labels or is the name enough to figure out where it came from?

Reply:
We might be a bit on the heavy side when it comes to labelling, we already set the owner.. But updating to do this for now and we can revisit in a broader discussion.